### PR TITLE
Consider evaporation on the interface velocity

### PR DESCRIPTION
**Pull request by mschreter**

### CODE REVIEW

**Comment by mschreter** [comment]
@peterrum would you be so kind and review this PR? Thanks.

**Comment by peterrum** [comment]
I'll do it after lunch. Quick question: is the evaporation working?

**Comment by mschreter** [comment]
So far I've only integrated evaporation in the level set problem (see animation -- sorry for the bad quality -- the interface moves due to evaporation)

anim

The next PR will consider evaporation + flow.

File: include/meltpooldg/utilities/amr.hpp [hunk 1]
```diff
@@ -18,93 +18,100 @@ namespace MeltPoolDG
...
+                        for (auto &face : cell->face_iterators())
+                          if (face->at_boundary())
+                            {
+                              if (cell->refine_flag_set())
+                                cell->clear_refine_flag();
+                              else
+                                cell->clear_coarsen_flag();
+                            }
+                      }
+                  }
```
Comment (peterrum) [inline include/meltpooldg/utilities/amr.hpp:66]: OK for now. But we should further investigate this!

File: include/meltpooldg/utilities/amr.hpp [hunk 2]
```diff
@@ -18,93 +18,100 @@ namespace MeltPoolDG
               const std::vector<std::pair<const DoFHandler<dim> *,
                                           std::function<void(std::vector<VectorType *> &)>>> &data,
               const std::function<void()> &                                                   post,
-              const std::function<void()> &setup_dof_system,
-              const AdaptiveMeshingData &  amr)
+              const std::function<void()> &      setup_dof_system,
+              const AdaptiveMeshingData<double> &amr,
```
Comment (peterrum) [inline include/meltpooldg/utilities/amr.hpp:22]: I think these changes makes sense: refining every n-th time step and limiting refinement is quite common! We should document this function at another time!

File: include/meltpooldg/level_set/level_set_operation.hpp [hunk 3]
```diff
@@ -540,7 +545,10 @@ namespace MeltPoolDG
       const LinearAlgebra::distributed::BlockVector<double> &
       get_normal_vector() const
       {
-        return reinit_operation->get_normal_vector();
+        if (level_set_data.do_reinitialization)
+          return reinit_operation->get_normal_vector();
+        else
+          return curvature_operation->get_normal_vector();
```
Comment (peterrum) [inline include/meltpooldg/level_set/level_set_operation.hpp:551]: Why is this needed? Why should one turn of reinit?

File: include/meltpooldg/interface/parameters.hpp [hunk 4]
```diff
@@ -37,14 +37,17 @@ namespace MeltPoolDG
...
-    bool         do_amr                      = false;
-    double       upper_perc_to_refine        = 0.0;
-    double       lower_perc_to_coarsen       = 0.0;
-    int          n_initial_refinement_cycles = 0;
-    unsigned int max_grid_refinement_level   = 12;
-    int          min_grid_refinement_level   = 1;
+    bool         do_amr                       = false;
+    bool         do_not_modify_boundary_cells = true;
+    number       upper_perc_to_refine         = 0.0;
+    number       lower_perc_to_coarsen        = 0.0;
```
Comment (peterrum) [inline include/meltpooldg/interface/parameters.hpp:46]: Introducing a template paramter `number` is not needed here.

File: include/meltpooldg/utilities/utilityfunctions.hpp [hunk 5]
```diff
@@ -96,8 +96,71 @@ namespace MeltPoolDG
       vec.compress(VectorOperation::max);
     }
 
+    /*
+     * @todo: replace Tensor<1, n_components, VectorizedArray> as template argument --> C++20
+     */
+    template <int dim, int n_components>
+    void
+    fill_dof_vector_from_cell_operation_vec(
```
Comment (peterrum) [inline include/meltpooldg/utilities/utilityfunctions.hpp:104]: ~How does this function differ from `fill_dof_vector_from_cell_operation`?~ Stupid question: the lambda ;)

File: include/meltpooldg/evaporation/evaporation_operation.hpp [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+  using namespace dealii;
+  /**
+   *     This module computes for a given evaporative mass flux the corresponding interface
+   *     velocity according to
+   *
+   *     \f[ \boldsymbol{n}\cfrac{\dot{m}}{\rho} \f]
+   *
+   *     with the normal vector \f$\boldsymbol{n}\f$, the evaporative mass flux \f$\dot{m}\f$
+   *     and the density \f$\rho\f$. One has to take care from which time step the normal
+   *     vector is computed
```
Comment (peterrum) [inline include/meltpooldg/evaporation/evaporation_operation.hpp:22]:
```suggestion
   *     vector is computed.
```

File: include/meltpooldg/evaporation/evaporation_operation.hpp [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+          ls.read_dof_values_plain(level_set);
+          ls.evaluate(true, false);
+
+          normal_vec.reinit(cell);
+          normal_vec.read_dof_values_plain(normal_vector);
+          normal_vec.evaluate(true, false);
+
+          for (unsigned int q_index = 0; q_index < ls.n_q_points; ++q_index)
+            {
+              //@todo: comment
```
Comment (peterrum) [inline include/meltpooldg/evaporation/evaporation_operation.hpp:87]: :) Indeed!

File: include/meltpooldg/evaporation/evaporation_operation.hpp [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+        [&](const unsigned int cell,
+            const unsigned int quad) -> const Tensor<1, dim, VectorizedArray<double>> & {
+          return begin_interface_velocity(cell)[quad];
+        });
+
+      scratch_data->get_constraint(vel_hanging_nodes_dof_idx).distribute(evaporation_velocity);
+    }
+
+    inline Tensor<1, dim, VectorizedArray<double>> *
+    begin_interface_velocity(const unsigned int macro_cell)
```
Comment (peterrum) [inline include/meltpooldg/evaporation/evaporation_operation.hpp:124]:
```suggestion
    begin_evaporation_velocity(const unsigned int macro_cell)
```

Maybe also define that is defined in the whole domain and not just on the "interface".

File: include/meltpooldg/evaporation/evaporation_operation.hpp [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+    {
+      AssertIndexRange(macro_cell, scratch_data->get_matrix_free().n_cell_batches());
+      AssertDimension(evaporation_velocities.size(),
+                      scratch_data->get_matrix_free().n_cell_batches() *
+                        scratch_data->get_matrix_free().get_n_q_points(ls_quad_idx) * dim);
+      return evaporation_velocities[scratch_data->get_matrix_free().get_n_q_points(ls_quad_idx) *
+                                    dim * macro_cell];
+    }
+
+  public:
```
Comment (peterrum) [inline include/meltpooldg/evaporation/evaporation_operation.hpp:145]:
```suggestion
```

File: include/meltpooldg/evaporation/evaporation_operation.hpp [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+
+    inline Tensor<1, dim, VectorizedArray<double>> *
+    begin_interface_velocity(const unsigned int macro_cell)
+    {
+      AssertIndexRange(macro_cell, scratch_data->get_matrix_free().n_cell_batches());
+      AssertDimension(evaporation_velocities.size(),
+                      scratch_data->get_matrix_free().n_cell_batches() *
+                        scratch_data->get_matrix_free().get_n_q_points(ls_quad_idx) * dim);
+      return &evaporation_velocities[scratch_data->get_matrix_free().get_n_q_points(ls_quad_idx) *
+                                     dim * macro_cell];
```
Comment (peterrum) [inline include/meltpooldg/evaporation/evaporation_operation.hpp:131]: `dim` is needed here because?

**Review by peterrum — APPROVED** [review]
Looks good to me. Some minor comments. 

A general remark: in the next weeks we should take the time and try to make dependencies of the operators more clear. I have lost a bit the overview.

File: include/meltpooldg/level_set/level_set_operation.hpp (hunk 3 above)
Comment (mschreter) [reply include/meltpooldg/level_set/level_set_operation.hpp:551]: One can play around by turning off reinitialization to investigate effects such as damping etc. inherent in the operations of reinitialization.

File: include/meltpooldg/utilities/utilityfunctions.hpp (hunk 5 above)
Comment (mschreter) [reply include/meltpooldg/utilities/utilityfunctions.hpp:104]: It would be nice to template the return type of the lambda function. However, this is a feature of C++20 which I did not investigate so far ;).

File: include/meltpooldg/simulations/stefans_problem/stefans_problem.hpp [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+  // const static bool isRegistered_1d =
+  // MeltPoolDG::SimulationFactory<1,1>::registerSimulation("stefans_problem",
+  // MeltPoolDG::makeDefaultSimulationFactoryFunction<SimulationStefansProblem<1>,1,1>() );
+  // const static bool isRegistered_2d = SimulationFactory<2>::registerSimulation("stefans_problem"
+  // makeDefaultSimulationFactoryFunction<class SimulationStefansProblem<2>>() );
+  // const static bool isRegistered_3d = SimulationFactory<3>::registerSimulation("stefans_problem"
+  // makeDefaultSimulationFactoryFunction<class SimulationStefansProblem<3>>() );
+  /*
+   * this function specifies the initial field of the level set equation
+   */
```
[the remark this answers is not in the document]
Comment (mschreter) [reply include/meltpooldg/simulations/stefans_problem/stefans_problem.hpp:229]: other topic ;)

File: include/meltpooldg/simulations/stefans_problem/stefans_problem.hpp [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+#include <deal.II/simplex/grid_generator.h>
+// c++
+#include <cmath>
+#include <iostream>
+// MeltPoolDG
+#include <meltpooldg/interface/simulationbase.hpp>
+#include <meltpooldg/simulations/simulation_factory.hpp>
+#include <meltpooldg/utilities/utilityfunctions.hpp>
+
+namespace MeltPoolDG::Simulation::StefansProblem
```
[the remark this answers is not in the document]
Comment (mschreter) [reply include/meltpooldg/simulations/stefans_problem/stefans_problem.hpp:22]: It is a first simple "toy" example. I will see what I can find ;-).

File: include/meltpooldg/evaporation/evaporation_operation.hpp (hunk 10 above)
Comment (mschreter) [reply include/meltpooldg/evaporation/evaporation_operation.hpp:131]: Since `evaporation_velocities` holds `Tensor<1,dim,VectorizedArray<double>>` objects?

Comment (peterrum) [reply include/meltpooldg/evaporation/evaporation_operation.hpp]: But are dim number of velocities per quadrature point stored?

Comment (mschreter) [reply include/meltpooldg/evaporation/evaporation_operation.hpp]: nope :-)

**Comment by mschreter** [comment]
All tests are passing (locally), thank you Peter for your careful comments.
